File: android/app/src/main/java/com/ayuvo/health/services/DailySummaryPolicy.kt
package com.ayuvo.health.services

enum class CalorieBalanceDirection {
    DEFICIT,
    SURPLUS,
    BALANCED
}

internal data class DailyCalorieBalance(
    val eatenCalories: Int,
    val burnedCalories: Int,
    val differenceCalories: Int,
    val direction: CalorieBalanceDirection
)

/** Pure nightly-summary math, separated from Android/Health Connect for JVM tests. */
internal object DailySummaryPolicy {
    fun resolveBurnedCalories(
        measuredTotalCalories: Int?,
        externalActiveCalories: Int,
        profileBmrCalories: Int?
    ): Int? {
        measuredTotalCalories?.takeIf { it > 0 }?.let { return it }
        val bmr = profileBmrCalories?.takeIf { it > 0 } ?: return null
        val active = externalActiveCalories.takeIf { it > 0 } ?: return null
        return bmr + active
    }

    fun balance(eatenCalories: Int, burnedCalories: Int): DailyCalorieBalance {
        val eaten = eatenCalories.coerceAtLeast(0)
        val burned = burnedCalories.coerceAtLeast(0)
        val signedDifference = burned - eaten
        val direction = when {
            signedDifference > 0 -> CalorieBalanceDirection.DEFICIT
            signedDifference < 0 -> CalorieBalanceDirection.SURPLUS
            else -> CalorieBalanceDirection.BALANCED
        }
        return DailyCalorieBalance(
            eatenCalories = eaten,
            burnedCalories = burned,
            differenceCalories = kotlin.math.abs(signedDifference),
            direction = direction
        )
    }
}

File: ios/calorietracker/Services/DailySummaryPolicy.swift
import Foundation

enum CalorieBalanceDirection: Equatable {
    case deficit
    case surplus
    case balanced
}

struct DailyCalorieBalance: Equatable {
    let eatenCalories: Int
    let burnedCalories: Int
    let differenceCalories: Int
    let direction: CalorieBalanceDirection
}

/// Pure home burn / nightly-summary math, separated from HealthKit for unit tests.
enum DailySummaryPolicy {
    static func resolveBurnedCalories(
        measuredTotalCalories: Int?,
        externalActiveCalories: Int,
        profileBmrCalories: Int?
    ) -> Int? {
        if let total = measuredTotalCalories, total > 0 { return total }
        guard let bmr = profileBmrCalories, bmr > 0 else { return nil }
        guard externalActiveCalories > 0 else { return nil }
        return bmr + externalActiveCalories
    }

    static func balance(eatenCalories: Int, burnedCalories: Int) -> DailyCalorieBalance {
        let eaten = max(eatenCalories, 0)
        let burned = max(burnedCalories, 0)
        let signedDifference = burned - eaten
        let direction: CalorieBalanceDirection
        if signedDifference > 0 {
            direction = .deficit
        } else if signedDifference < 0 {
            direction = .surplus
        } else {
            direction = .balanced
        }
        return DailyCalorieBalance(
            eatenCalories: eaten,
            burnedCalories: burned,
            differenceCalories: abs(signedDifference),
            direction: direction
        )
    }
}
